### PR TITLE
Added buffers in breakpoints for secagg

### PR DESCRIPTION
**Pull request by lchambon**
**PR description**

#1003 
This PR introduces persistent buffers parameters in breakpoints during secure aggregation. 
It also fixes some issues with model weight API, (#1010 ) by adding the only_trainable and exclude_buffers boolean in flatten and unflatten functions.

**Developer Certificate Of Origin (DCO)**

By opening this merge request, you agree to the
[Developer Certificate of Origin (DCO)](https://github.com/fedbiomed/fedbiomed/blob/master/CONTRIBUTING.md#fed-biomed-developer-certificate-of-origin-dco)

This DCO essentially means that:

- you offer the changes under the same license agreement as the project, and
- you have the right to do that,
- you did not steal somebody else’s work.

**License**

Project code files should begin with these comment lines to help trace their origin:
```
# This file is originally part of Fed-BioMed
# SPDX-License-Identifier: Apache-2.0
```

Code files can be reused from another project with a compatible non-contaminating license.
They shall retain the original license and copyright mentions.
The `CREDIT.md` file and `credit/` directory shall be completed and updated accordingly.

**Guidelines for PR review**

General:

* give a glance to [DoD](https://fedbiomed.org/latest/developer/definition-of-done/)
* check [coding rules and coding style](https://fedbiomed.org/latest/developer/usage_and_tools/#coding-style)
* check docstrings (eg run `tests/docstrings/check_docstrings`)

Specific to some cases:

* update all conda envs consistently (`development` and `vpn`, Linux and MacOS)
* if modified researcher (eg new attributes in classes) check if breakpoint needs update (`breakpoint`/`load_breakpoint` in `Experiment()`, `save_state_breakpoint`/`load_state_breakpoint` in aggregators, strategies, secagg, etc.)
* if modified a component with versioning (config files, breakpoint, messaging protocol) then update the version following the rules in `common/utils/_versions.py`

### CODE REVIEW

**Comment by mvesin** [comment]
hi @lchambon @sharkovsky 

I looked for remaining calls to `model.parameters()` (except in unit tests):

1. here https://github.com/fedbiomed/fedbiomed/blob/49bfaf6b5d6f5474fdebfdfaeff18d38c7dcf7ad/fedbiomed/common/training_plans/_torchnn.py#L215 shouldn't we be calling rather `get_weights()` than `model.parameters()` for the case of a model with non-trainable parameters to optimize (don't know if this makes sense though ...)

```
self._optimizer = torch.optim.Adam(self._model.get_weights(only_trainable=only_trainable, exclude_buffers=exclude_buffers).values(), **self._optimizer_args)
```

Doc suggests Adam optimizer can optimizer a full dict
https://pytorch.org/docs/1.13/generated/torch.optim.Adam.html?highlight=torch+optim+adam#torch.optim.Adam

Yet I did not check if `only_trainable` and `exclude_buffers` can easily be made available here.

2. misc documentation. Maybe also update this https://github.com/fedbiomed/fedbiomed/blob/49bfaf6b5d6f5474fdebfdfaeff18d38c7dcf7ad/fedbiomed/common/optimizers/generic_optimizers.py#L458 to

```
    >>> optimizer = torch.optim.SGD(model.get_weights(), .1)
```

... to give good habit of using get_weights

**Comment by mvesin** [comment]
@lchambon @sharkovsky regarding `Model.unflatten()` here https://github.com/fedbiomed/fedbiomed/blob/49bfaf6b5d6f5474fdebfdfaeff18d38c7dcf7ad/fedbiomed/common/models/_model.py#L189

Maybe we want to change prototype to the following for API coherence (same parameters for parent & child class, as for `flatten()`)

```
    def unflatten(
            self,
            weights_vector: List[float],
            only_trainable: bool = False,
            exclude_buffers: bool = True
    ) -> None:
```

Not very big deal anyway ...

**Comment by sharkovsky** [comment]
> hi @lchambon @sharkovsky
> 
> I looked for remaining calls to `model.parameters()` (except in unit tests):
> 

> 


Hi Marc, 

thanks for double-checking!

For your first point:
>     1. here https://github.com/fedbiomed/fedbiomed/blob/49bfaf6b5d6f5474fdebfdfaeff18d38c7dcf7ad/fedbiomed/common/training_plans/_torchnn.py#L215
>         shouldn't we be calling rather `get_weights()` than `model.parameters()` for the case of a model with non-trainable parameters to optimize (don't know if this makes sense though ...)
> 
> 
> ```
> self._optimizer = torch.optim.Adam(self._model.get_weights(only_trainable=only_trainable, exclude_buffers=exclude_buffers).values(), **self._optimizer_args)
> ```
> 
> Doc suggests Adam optimizer can optimizer a full dict https://pytorch.org/docs/1.13/generated/torch.optim.Adam.html?highlight=torch+optim+adam#torch.optim.Adam
> 
> Yet I did not check if `only_trainable` and `exclude_buffers` can easily be made available here.

The short answer is no. In Pytorch we explicitly pass **only the trainable parameters** to the optimizer, because the buffers are, by definition, not trainable. 
Alternatively, we could convert it to `get_weights` in the following manner:
```python
self._optimizer = torch.optim.Adam(self.model().get_weights(only_trainable=True, exclude_buffers=True).values(), **self._optimizer_args)
```

For your second point
>     2. misc documentation. Maybe also update this https://github.com/fedbiomed/fedbiomed/blob/49bfaf6b5d6f5474fdebfdfaeff18d38c7dcf7ad/fedbiomed/common/optimizers/generic_optimizers.py#L458
>         to
> 
> 
> ```
>     >>> optimizer = torch.optim.SGD(model.get_weights(), .1)
> ```
> 
> ... to give good habit of using get_weights

I'm also not sure about this: I think we want to keep the user-facing code as similar as possible to PyTorch, which is also the original reason why `TrainingPlan.model()` returns the underlying pytorch model, and not the wrapper object. 

What do you think?

**Review by mvesin — APPROVED** [review]
Hi @sharkovsky @lchambon 

Once we have concluded on the small point regarding `Model.unflatten()` prototype, this is OK for me to merge.

Thank you very much for this nice work (though one !)

**Comment by sharkovsky** [comment]
> @lchambon @sharkovsky regarding `Model.unflatten()` here
> 
> https://github.com/fedbiomed/fedbiomed/blob/49bfaf6b5d6f5474fdebfdfaeff18d38c7dcf7ad/fedbiomed/common/models/_model.py#L189
> 
> Maybe we want to change prototype to the following for API coherence (same parameters for parent & child class, as for `flatten()`)
> 
> ```
>     def unflatten(
>             self,
>             weights_vector: List[float],
>             only_trainable: bool = False,
>             exclude_buffers: bool = True
>     ) -> None:
> ```
> 
> Not very big deal anyway ...

Sounds good to me, it would be nicer to change the base (abstract) method signature. 

@lchambon could you please submit a commit for that?

**Comment by lchambon** [comment]
Hi @sharkovsky @mvesin 
yes sure, I will take care of changing the prototype for unflatten.

File: fedbiomed/common/models/_sklearn.py [hunk 1]
```diff
@@ -165,12 +165,18 @@ def flatten(self,
...
-            weights_vector: List[float]
+            weights_vector: List[float],
+            only_trainable: bool = False,
+            exclude_buffers: bool = True
     ) -> Dict[str, np.ndarray]:
         """Unflatten vectorized model weights
 
         Args:
             weights_vector: Vectorized model weights to convert dict
+            only_trainable: Unused for scikit-learn models. (Whether to ignore
```
Comment (mvesin) [inline fedbiomed/common/models/_sklearn.py:176]: dummy

**Comment by lchambon** [comment]
Hi @sharkovsky @mvesin, I updated the signature of the unflatten abstract method.

**Pull request merged**